common/usermode.c

```diff
@@ -49,8 +49,13 @@ static inline void syscall_restore(void) {
 }
 
 static inline long syscall_return(long return_code) {
-    register long ret asm(STR(_ASM_AX)) = return_code;
-    return ret;
+    /* clang-format off */
+    asm volatile(
+            "mov %[ret], %%" STR(_ASM_AX) "\n"
+            ::[ret] "r"(return_code)
+    );
+    /* clang-format on */
+    return return_code;
 }
```

Comment:
WTF is this?! Why isn't this a simple `return return_code`?.... Oh, I see. `syscall_handler()`, what a bummer! It does the right thing only by chance. We'd better rewrite that think with a ASM thunk instead of this fragile register pinning code (gcc is free to reuse these registers, you know ;)).

Reply:
You are right of course, the current implementation of syscall handling is indeed quite fragile. I plan to improve it all at some point.